File: frontend/src/lib/api/audit-logs.ts
import { apiClient } from "./client";
import type { AuditLogFilter, PaginatedAuditLogsResponse } from "./types";

export const auditLogsApi = {
  list: (params?: { page?: number; page_size?: number } & AuditLogFilter) => {
    const searchParams = new URLSearchParams();
    if (params?.page) searchParams.set("page", params.page.toString());
    if (params?.page_size) searchParams.set("page_size", params.page_size.toString());
    if (params?.action) searchParams.set("action", params.action);
    if (params?.entity_type) searchParams.set("entity_type", params.entity_type);
    if (params?.status) searchParams.set("status", params.status);
    if (params?.start_date) searchParams.set("start_date", params.start_date);
    if (params?.end_date) searchParams.set("end_date", params.end_date);
    const query = searchParams.toString();
    return apiClient.get<PaginatedAuditLogsResponse>(`/audit-logs${query ? `?${query}` : ""}`);
  },
};

File: backend/src/api/handlers/audit_logs.rs
use std::sync::Arc;

use axum::{
    extract::{Query, State},
    Json,
};

use crate::api::extractors::{AuthUser, PaginatedResponse, Pagination};
use crate::domain::models::{AuditLogFilter, AuditLogResponse};
use crate::domain::services::AuditLogService;
use crate::error::AppResult;

pub type AuditLogServiceState = Arc<AuditLogService>;

#[utoipa::path(
    get,
    path = "/api/v1/audit-logs",
    params(
        ("page" = Option<i64>, Query, description = "Page number (default: 1)"),
        ("page_size" = Option<i64>, Query, description = "Items per page (default: 20)"),
        ("action" = Option<String>, Query, description = "Filter by action"),
        ("entity_type" = Option<String>, Query, description = "Filter by entity type"),
        ("status" = Option<String>, Query, description = "Filter by status"),
        ("start_date" = Option<String>, Query, description = "Filter by start date"),
        ("end_date" = Option<String>, Query, description = "Filter by end date")
    ),
    responses(
        (status = 200, description = "List of audit logs with pagination"),
        (status = 401, description = "Unauthorized")
    ),
    tag = "Audit Logs",
    security(("bearer" = []))
)]
pub async fn list_audit_logs(
    State(audit_service): State<AuditLogServiceState>,
    auth_user: AuthUser,
    pagination: Pagination,
    Query(filter): Query<AuditLogFilter>,
) -> AppResult<Json<PaginatedResponse<AuditLogResponse>>> {
    let logs = audit_service
        .list(
            auth_user.id(),
            auth_user.is_admin(),
            &filter,
            pagination.limit,
            pagination.offset,
        )
        .await?;

    let total = audit_service
        .count(auth_user.id(), auth_user.is_admin(), &filter)
        .await?;

    Ok(Json(PaginatedResponse::new(logs, &pagination, total)))
}
